electrum/plugins/keepkey/keepkey.py

```diff
@@ -198,52 +196,8 @@ def get_client(self, keystore, force_pair=True, *,
     def get_coin_name(self):
         return "Testnet" if constants.net.TESTNET else "Bitcoin"
 
-    def initialize_device(self, device_id, wizard, handler):
-        # Initialization method
-        msg = _("Choose how you want to initialize your {}.\n\n"
-                "The first two methods are secure as no secret information "
-                "is entered into your computer.\n\n"
-                "For the last two methods you input secrets on your keyboard "
-                "and upload them to your {}, and so you should "
-                "only do those on a computer you know to be trustworthy "
-                "and free of malware."
-        ).format(self.device, self.device)
-        choices = [
-            # Must be short as QT doesn't word-wrap radio button text
-            (TIM_NEW, _("Let the device generate a completely new seed randomly")),
-            (TIM_RECOVER, _("Recover from a seed you have previously written down")),
-            (TIM_MNEMONIC, _("Upload a BIP39 mnemonic to generate the seed")),
-            (TIM_PRIVKEY, _("Upload a master private key"))
-        ]
-        def f(method):
-            import threading
-            settings = self.request_trezor_init_settings(wizard, method, self.device)
-            t = threading.Thread(target=self._initialize_device_safe, args=(settings, method, device_id, wizard, handler))
-            t.daemon = True
-            t.start()
-            exit_code = wizard.loop.exec_()
-            if exit_code != 0:
-                # this method (initialize_device) was called with the expectation
-                # of leaving the device in an initialized state when finishing.
-                # signal that this is not the case:
-                raise UserCancelled()
-        wizard.choice_dialog(title=_('Initialize Device'), message=msg, choices=choices, run_next=f)
-
-    def _initialize_device_safe(self, settings, method, device_id, wizard, handler):
-        exit_code = 0
-        try:
-            self._initialize_device(settings, method, device_id, wizard, handler)
-        except UserCancelled:
-            exit_code = 1
-        except BaseException as e:
-            self.logger.exception('')
-            handler.show_error(repr(e))
-            exit_code = 1
-        finally:
-            wizard.loop.exit(exit_code)
-
     @runs_in_hwd_thread
-    def _initialize_device(self, settings, method, device_id, wizard, handler):
+    def _initialize_device(self, settings, method, device_id, handler):
```

Comment:
Restoring with TIM_RECOVER is broken. The value of `word_count` being passed used to be 24, now it is 6 (item is `-1` now). The value does not matter, but it has to be one of {12, 18, 24}.
```patch
diff --git a/electrum/plugins/keepkey/keepkey.py b/electrum/plugins/keepkey/keepkey.py
index ecbb0444e8..5d81b13a37 100644
--- a/electrum/plugins/keepkey/keepkey.py
+++ b/electrum/plugins/keepkey/keepkey.py
@@ -211,7 +211,7 @@ class KeepKeyPlugin(HW_PluginBase):
             client.reset_device(True, strength, passphrase_protection,
                                 pin_protection, label, language)
         elif method == TIM_RECOVER:
-            word_count = 6 * (item + 2)  # 12, 18 or 24
+            word_count = 24  # looks like this value is ignored by the device, but it has to be one of {12,18,24}
             client.step = 0
             client.recovery_device(word_count, passphrase_protection,
                                        pin_protection, label, language)

```

Reply:
This might have already been broken. The changes in this PR don't really change the handling of the radiobuttons

Reply:
No, the bug is introduced with this PR.


https://github.com/spesmilo/electrum/blob/44a15951575254572f49d81f189cd4456b488815/electrum/plugins/keepkey/qt.py#L304
`bg.checkedId()`, before the PR, is called in the same method that creates `bg`. After the PR, it is called in a separate method.
Before the PR, it returns `2`. After the PR, it returns `-1`.
I think it is the groupbox `gb`, that gets GC-ed when we go out of scope, because of lack of parenting of objects:
https://github.com/spesmilo/electrum/blob/44a15951575254572f49d81f189cd4456b488815/electrum/plugins/keepkey/qt.py#L246

E.g. with this patch, the issue goes away:
```patch
diff --git a/electrum/plugins/keepkey/qt.py b/electrum/plugins/keepkey/qt.py
index 014cce74c8..efa0dc1255 100644
--- a/electrum/plugins/keepkey/qt.py
+++ b/electrum/plugins/keepkey/qt.py
@@ -241,9 +241,10 @@ class KeepkeyInitLayout(QVBoxLayout):
         hl.addWidget(self.label_e)
         hl.addStretch(1)
         self.addLayout(hl)
+        self._w = QWidget()
 
         if self.method in [TIM_NEW, TIM_RECOVER]:
-            gb = QGroupBox()
+            gb = QGroupBox(self._w)
             hbox1 = QHBoxLayout()
             gb.setLayout(hbox1)
             # KeepKey recovery doesn't need a word count
@@ -260,6 +261,7 @@ class KeepkeyInitLayout(QVBoxLayout):
                 rb.setChecked(True)
             self.cb_pin = QCheckBox(_('Enable PIN protection'))
             self.cb_pin.setChecked(True)
+            print(f"heyheyhey1. {self.bg.checkedId()=!r}")
         else:
             self.text_e = QTextEdit()
             self.text_e.setMaximumHeight(60)
@@ -301,6 +303,7 @@ class KeepkeyInitLayout(QVBoxLayout):
 
     def get_settings(self):
         if self.method in [TIM_NEW, TIM_RECOVER]:
+            print(f"heyheyhey2. {self.bg.checkedId()=!r}")
             item = self.bg.checkedId()
             pin = self.cb_pin.isChecked()
         else:
```

Only TIM_RECOVER is affected. TIM_NEW is not, as `gb` there won't get GC-ed:
https://github.com/spesmilo/electrum/blob/44a15951575254572f49d81f189cd4456b488815/electrum/plugins/keepkey/qt.py#L250-L251

Reply:
I've pushed the first patch in https://github.com/spesmilo/electrum/commit/0a3dd8e5e53e9a95480cf1700d6ba2a0008f9a37